File: src/main/java/com/derongan/minecraft/mineinabyss/player/PlayerDataImpl.java
package com.derongan.minecraft.mineinabyss.player;

import com.derongan.minecraft.mineinabyss.ascension.effect.AscensionEffect;
import com.derongan.minecraft.mineinabyss.whistles.WhistleType;
import com.derongan.minecraft.mineinabyss.world.Layer;
import org.bukkit.entity.Player;

import java.util.Date;
import java.util.List;


public class PlayerDataImpl implements PlayerData {
    private Layer currentLayer;

    private boolean affectedByCurse;
    private boolean anchored;
    private boolean ingame = false;

    private double distanceAscended;
    private double exp;
    private double expOnDescent;

    private Player player;
    private WhistleType whistle = WhistleType.BELL;
    private Date descentDate;
    private List<AscensionEffect> effects;

    public PlayerDataImpl(Player player) {
        this.player = player;
        this.affectedByCurse = true;
    }

    @Override
    public double getExpOnDescent() {
        return expOnDescent;
    }

    @Override
    public void setExpOnDescent(double expOnDescent) {
        this.expOnDescent = expOnDescent;
    }

    @Override
    public Date getDescentDate() {
        return descentDate;
    }

    @Override
    public void setDescentDate(Date date) {
        descentDate = date;
    }

    @Override
    public boolean isIngame() {
        return ingame;
    }

    @Override
    public void setIngame(boolean ingame) {
        this.ingame = ingame;
    }

    @Override
    public Layer getCurrentLayer() {
        return currentLayer;
    }

    @Override
    public void setCurrentLayer(Layer currentLayer) {
        this.currentLayer = currentLayer;
    }

    @Override
    public boolean isAffectedByCurse() {
        return affectedByCurse;
    }

    @Override
    public void setAffectedByCurse(boolean affectedByCurse) {
        this.affectedByCurse = affectedByCurse;
    }

    @Override
    public boolean isAnchored() {
        return anchored;
    }

    @Override
    public void setAnchored(boolean anchored) {
        this.anchored = anchored;
    }

    @Override
    public Player getPlayer() {
        return player;
    }

    @Override
    public List<AscensionEffect> getAscensionEffects() {
        return effects;
    }

    @Override
    public void addAscensionEffect(AscensionEffect effect) {
        effects.add(effect);
    }

    @Override
    public double getDistanceAscended() {
        return distanceAscended;
    }

    @Override
    public void setDistanceAscended(double distanceAscended) {
        this.distanceAscended = distanceAscended;
    }

    @Override
    public WhistleType getWhistle() {
        return whistle;
    }

    @Override
    public void setWhistle(WhistleType whistle) {
        this.whistle = whistle;
    }

    @Override
    public int getLevel() {
        return (int) exp / 10; //TODO write a proper formula
    }

    @Override
    public double getExp() {
        return exp;
    }

    @Override
    public void setExp(double exp) {
        this.exp = exp;
    }

    @Override
    public void addExp(double exp) {
        this.exp += exp;
    }
}

File: src/main/java/com/derongan/minecraft/mineinabyss/player/PlayerDataImpl.kt
package com.derongan.minecraft.mineinabyss.player

import com.derongan.minecraft.mineinabyss.ascension.effect.AscensionEffect
import com.derongan.minecraft.mineinabyss.world.Layer
import com.mineinabyss.idofront.serialization.UUIDSerializer
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*

@Serializable
class PlayerDataImpl(
    @Serializable(with = UUIDSerializer::class)
    val uuid: UUID, //TODO pass this from file name somehow
    @SerialName("affectable")
    override var isAffectedByCurse: Boolean = true,
    @SerialName("ingame")
    override var isIngame: Boolean = false,
    @SerialName("ascended")
    override var curseAccrued: Double = 0.0,
    override var exp: Double = 0.0,
    override var expOnDescent: Double = 0.0,
    @Serializable(with = DateAsLongSerializer::class)
    override var descentDate: Date? = null,
) : PlayerData {
    @Transient
    override var currentLayer: Layer? = null

    @Transient
    override val player: Player = Bukkit.getPlayer(uuid) ?: error("UUID of player data doesn't match a player")

    @Transient
    private val effects: MutableList<AscensionEffect> = mutableListOf()
    override val ascensionEffects get() = effects.toList()
    override val level: Int get() = exp.toInt() / 10 //TODO write a proper formula

    override fun addAscensionEffect(effect: AscensionEffect) {
        effects.add(effect)
    }

    override fun addExp(exp: Double) {
        this.exp += exp
    }
}

object DateAsLongSerializer : KSerializer<Date> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Date", PrimitiveKind.LONG)
    override fun serialize(encoder: Encoder, value: Date) = encoder.encodeLong(value.time)
    override fun deserialize(decoder: Decoder): Date = Date(decoder.decodeLong())
}
